pkg/resources/envoy/envoy.go

```diff
@@ -53,41 +71,87 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
+		if r.KafkaCluster.Spec.GetIngressController() == envoyutils.IngressControllerName && eListener.GetAccessMethod() == corev1.ServiceTypeLoadBalancer {
+			ingressConfigs, defaultControllerName, err := util.GetIngressConfigs(r.KafkaCluster.Spec, eListener)
+			if err != nil {
+				return err
+			}
+			var externalListenerResources []resources.ResourceWithLogAndExternalListenerSpecificInfos
+			externalListenerResources = append(externalListenerResources,
+				r.service,
+				r.configMap,
+				r.deployment,
+			)
+
+			if r.KafkaCluster.Spec.EnvoyConfig.GetDistruptionBudget().DisruptionBudget.Create {
+				externalListenerResources = append(externalListenerResources, r.podDisruptionBudget)
+			}
+			for name, ingressConfig := range ingressConfigs {
+				if !util.IsIngressConfigInUse(name, defaultControllerName, r.KafkaCluster, log) {
+					continue
+				}
 
-	if r.KafkaCluster.Spec.GetIngressController() == envoyutils.IngressControllerName {
-		for _, eListener := range r.KafkaCluster.Spec.ListenersConfig.ExternalListeners {
-			if eListener.GetAccessMethod() == corev1.ServiceTypeLoadBalancer {
-				ingressConfigs, defaultControllerName, err := util.GetIngressConfigs(r.KafkaCluster.Spec, eListener)
-				if err != nil {
-					return err
+				for _, res := range externalListenerResources {
+					o := res(log, eListener, ingressConfig, name, defaultControllerName)
+					err := k8sutil.Reconcile(log, r.Client, o, r.KafkaCluster)
+					if err != nil {
+						return err
+					}
 				}
-				var externalListenerResources []resources.ResourceWithLogAndExternalListenerSpecificInfos
-				externalListenerResources = append(externalListenerResources,
-					r.service,
-					r.configMap,
-					r.deployment,
-				)
-
-				if r.KafkaCluster.Spec.EnvoyConfig.GetDistruptionBudget().DisruptionBudget.Create {
-					externalListenerResources = append(externalListenerResources, r.podDisruptionBudget)
+			}
+		} else {
+			// Cleaning up unused envoy resources when ingress controller is not envoy or eListener access method is not LoadBalancer
+			deletionCounter := 0
+			ctx := context.Background()
+			envoyResourcesGVK := []schema.GroupVersionKind{
+				{
+					Version: corev1.SchemeGroupVersion.Version,
+					Group:   corev1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(corev1.Service{}).Name(),
+				},
+				{
+					Version: corev1.SchemeGroupVersion.Version,
+					Group:   corev1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(corev1.ConfigMap{}).Name(),
+				},
+				{
+					Version: appsv1.SchemeGroupVersion.Version,
+					Group:   appsv1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(appsv1.Deployment{}).Name(),
+				},
+				{
+					Version: policyv1.SchemeGroupVersion.Version,
+					Group:   policyv1.SchemeGroupVersion.Group,
+					Kind:    reflect.TypeOf(policyv1.PodDisruptionBudget{}).Name(),
+				},
+			}
+			var envoyResources unstructured.UnstructuredList
```

Comment:
Do we need to define this outside of the cycle? As far as I can tell it's only used inside.

Reply:
I used it outside to be able to re-use it inside iteration but I think it does not matter much.